dockers/docker-sonic-telemetry/certificate_rotation_checker

```diff
@@ -0,0 +1,225 @@
+                              "Telemetry server process is running after certificate and private key were rotated!")
+                break
+
+    if not is_running:
+        syslog.syslog(syslog.LOG_ERR,
+                      "Telemetry server process is not running after certificate and private key were rotated and exiting ...")
+        sys.exit(1)
+
+
+def restart_telemetry_server():
+    """Restarts the telemetry server process by Supervisord and then checks
+    it is actually running.
+
+    Args:
+        None
+
+    Returns:
+        None
+    """
+    restart_telemetry_server_cmd = "supervisorctl restart telemetry"
+    retry_times = 0
+
+    while retry_times <= MAX_RETRY_TIMES:
+        retry_times += 1
+        exit_code, command_stdout = get_command_result(restart_telemetry_server_cmd)
+        if exit_code != 0:
+            syslog.syslog(syslog.LOG_INFO,
+                          "Failed to restart telemetry server process and retry after 60 seconds ...")
+            time.sleep(60)
+        else:
+            break
+
+    if retry_times > MAX_RETRY_TIMES:
+        syslog.syslog(syslog.LOG_ERR,
+                      "Failed to restart telemetry server process after trying '{}' times and exiting ..."
+                      .format(MAX_RETRY_TIMES))
+        sys.exit(2)
+
+    check_telemetry_server_running()
+
+
+def check_certificate_rotated(certificate_file_name, private_key_file_name):
+    """Leverages the 'inotify' module to monitor the file system events under the
+    directory which stores the SONiC credentials and restarts telemetry server
+    process if certificate and private key were rotated.
+
+
+    Args:
+        certificate_file_name: A string indicates the telemetry certificate file name.
+        private_key_file_name: A string indicates the telemetry private key file name.
+
+    Returns:
+        None.
+    """
+    certificate_file_rotated = False
+    private_key_file_rotated = False
+
+    inotify_instance = inotify.adapters.Inotify()
+    inotify_instance.add_watch(CREDENTIALS_DIR_PATH)
+    for event in inotify_instance.event_gen(yield_nones=False):
+        header, event_type, monitoring_path, file_name = event
+        if (file_name == certificate_file_name
+                and ("IN_CREATE" in event_type or "IN_MOVED_TO" in event_type)):
+            certificate_file_rotated = True
+        if (file_name == private_key_file_name
+                and ("IN_CREATE" in event_type or "IN_MOVED_TO" in event_type)):
+            private_key_file_rotated = True
+
+        if certificate_file_rotated and private_key_file_rotated:
+            certificate_file_rotated = False
+            private_key_file_rotated = False
+            syslog.syslog(syslog.LOG_INFO,
+                          "Certificate and private key were rotated and restarting telemetry server process ...")
+            restart_telemetry_server()
+
+            # Wait for specified seconds and then do the next round checking
```

Comment:
>Wait for specified second [](http://example.com/codeflow?start=14&length=25)

You don't need sleep in an event loop.

Reply:
Initially what I am thinking is since the directory `/etc/sonic/credentials/` will store all the certificates and keys for different applications in the future and if one or several certificates under this directory are rotated too frequently (by accident?), this event loop will be executed continuously and then burn too much CPU cycles potentially.  Another thinking the polling frequency of our internal tool to get the update from certificate storage is 8 ~ 24 hours. That's why I put a sleep function at here.

I will let `inotify` to monitor the certificate file directly instead of monitoring the `/etc/sonic/credentials/` and then we can remove this sleep function in event loop.

Reply:
Got your motivation, and the considering may be helpful. However, event you sleep, the event will be queued and processed after each sleep, you don't save too much?

Reply:
Agree and this will not save too much!